### PR TITLE
Re-add token-based and contact-based recovery

### PR DESCRIPTION
**Pull request by bifurcation**
This PR re-adds contact- and token-based recovery in ways that are (1) MitM resistant, and (2) more compatible with RESTification.  As a side-effect, it adds a key agreement method that we will probably re-use in the redesigned DNS challenge.

### CODE REVIEW

File: draft-barnes-acme.md [hunk 1]
```diff
@@ -516,18 +544,211 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+
+In addition to the recovery mechanisms defined by ACME, individual client implementations may also offer implementation-specific recovery mechanisms.  For example, if a client creates account keys deterministically from a seed value, then this seed could be used to recover the account key by re-generating it.  Or an implementation could escrow an encrypted copy of the account key with a cloud storage provider, and give the encryption key to the user as a recovery value.
+
+### Token-Based Recovery
+
+With token-based recovery, the client proves to the server that it holds a secret value established in the initial registration transaction.  The client requests token-based recovery by sending a MAC over the new account key, using the recovery key from the initial registration.
+
+method (required, string):
+: The string "contact"
```
Comment (jcjones) [inline draft-barnes-acme.md]: token, I thin?

File: draft-barnes-acme.md [hunk 2]
```diff
@@ -516,18 +544,211 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+: The client's ECDH public key
+
+server (optional, JWK):
+: The server's ECDH public key
+
+length (optional, number):
+: The length, in bits, of the derived secret.  The value MUST be a multiple of 8.
+
+confirm (optional, string):
+: The confirmation value, base64-encoded
```
Comment (jcjones) [inline draft-barnes-acme.md]: The use of the confirmation value is unspecified, and X9.63 doesn't define the use either. Is it an HMAC? Is it implementation-specific?

File: draft-barnes-acme.md [hunk 3]
```diff
@@ -516,18 +544,211 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+* The MAC verifies using the recovery key
+* The JWK in the payload represents the new account key (i.e. the key used to verify the ACME message)
+
+If those conditions are met, and the recovery request is otherwise acceptable to the server, then the recovery process has succeeded.  The server creates a new registration resource based on the base registration and the new account key, and returns it on a 201 (Created) response, together with a Location header indicating a URI for the new registration.  If the recovery request is unsuccessful, the server returns an error response, such as 403 (Forbidden).
+
+~~~~~~~~~~
 
+HTTP/1.1 201 Created
+Content-Type: application/json
+Location: https://example.com/reg/asdf
```
Comment (jcjones) [inline draft-barnes-acme.md]: nit: missing the `/acme`

File: draft-barnes-acme.md [hunk 4]
```diff
@@ -516,18 +544,211 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+
+~~~~~~~~~~
+
+If the server agrees to attempt contact-based recovery, then it creates a new registration resource containing a stub registration object.  The stub registration has the client's new account key and anonymized contacts, in order to allow the the client to know which contacts to check.  The server returns the stub contact in a 201 (Created) response, along with a Location header field indicating the URI for the new registration resource (which will be the registration URI if the recovery succeeds).
+
+~~~~~~~~~~
+
+HTTP/1.1 201 Created
+Content-Type: application/json
+Location: https://example.com/reg/qwer
```
Comment (jcjones) [inline draft-barnes-acme.md]: nit: Also missing `/acme`

Comment (jcjones) [inline draft-barnes-acme.md]: Actually, this is true in a variety of places. Maybe I'm just ignorant about the difference in URLs.

File: draft-barnes-acme.md (hunk 2 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: The idea is that the client could use this to verify that it has gotten the right derived key, e.g., that a MitM hasn't set length to be really short.

Comment (jcjones) [inline draft-barnes-acme.md]: Ah. I read this section as being a value that you could use to determine whether you had correctly derived the same shared secret, to confirm that the recovery will work in the future.

File: draft-barnes-acme.md (hunk 4 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: It doesn't actually matter, because the URLs are examples.  But I fixed the other ones.

File: draft-barnes-acme.md [hunk 5]
```diff
@@ -427,6 +435,24 @@ of {{RFC7515}}.  If the value of a "nonce" header parameter is not
...
 
+## Key Agreement
+
+Certain elements of the protocol will require the establishment of a shared secret between the client and the server, in such a way that an entity observing the ACME protocol cannot derive the secret.  In these cases, we use a simple ECDH key exchange, based on the system used by CMS {{RFC5753}}:
+
+* Inputs:
+  * Client-generated key pair
+  * Server-generated key pair
+  * Length of the shared secret to be derived
+* Perform the ECDH primitive operation to obtain X_P (Section 3.3.1 of {{SEC1}})
```
Comment (ekr) [inline draft-barnes-acme.md]: ITYM Z

File: draft-barnes-acme.md [hunk 6]
```diff
@@ -427,6 +435,24 @@ of {{RFC7515}}.  If the value of a "nonce" header parameter is not
...
+* Select a hash algorithm according to the curve being used:
+  * For "P-256", use SHA-256
+  * For "P-384", use SHA-384
+  * For "P-512", use SHA-512
+* Derive the shared secret value using the KDF in Section 3.6.1 of {{SEC1}} using the selected hash algorithm
```
Comment (ekr) [inline draft-barnes-acme.md]: It seems like you're really burying the lede here, since the point of this exercise is to generate
a key which is << 256 bits, but you don't explain how to do that.

File: draft-barnes-acme.md [hunk 7]
```diff
@@ -427,6 +435,24 @@ of {{RFC7515}}.  If the value of a "nonce" header parameter is not
...
+
+The parties may also generate a key confirmation value H(ZZ), where ZZ is the shared secret and H is the selected hash algorithm.  Use of the confirmation value prevents tampering with protocol elements that encode the length of the desired secret, but allow an attacker to attempt to recover the shared secret via a preimage attack.
+
```
Comment (ekr) [inline draft-barnes-acme.md]: I can't read this text.

File: draft-barnes-acme.md [hunk 8]
```diff
@@ -516,18 +544,234 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
 
 If the client wishes to update this information in the future, it sends a POST request with updated information to the registration URI.  The server MUST ignore any updates to the "key", "authorizations, or "certificates" fields, and MUST verify that the request is signed with the private key corresponding to the "key" field of the request before updating the registration.
 
-Servers SHOULD NOT respond to GET requests for registration resources as these requests are not authenticated.  If a client wishes to query the server for information about its account (e.g., to examine the "contact" or "certificates" fields), then it SHOULD do so by sending a POST request with an empty update.  That is, it should send a JWS whose payload is the empty JSON dictionary ("{}").
+Servers SHOULD NOT respond to GET requests for registration resources as these requests are not authenticated.  If a client wishes to query the server for information about its account (e.g., to examine the "contact" or "certificates" fields), then it SHOULD do so by sending a POST request with an empty update.  That is, it should send a JWS whose payload is trivial ({"resource":"reg"}).
+
```
Comment (ekr) [inline draft-barnes-acme.md]: This seems like it should be in a separate PR.

File: draft-barnes-acme.md [hunk 9]
```diff
@@ -516,18 +544,234 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+If the client wishes to establish a secret key with the server that it can use to recover this account later (a "recovery key"), then it must perform a simple key agreement protocol as part of the new-registration transaction.  The client and server perform an ECDH exchange through the new-registration transaction, and the result is the recovery key.
+
+To request a recovery key, the client includes a "recoveryKey" field in its new-registration request.  The value of this field is a JSON object.
+
+client (required, JWK):
+: The client's ECDH public key
 
+server (optional, JWK):
+: The server's ECDH public key
```
Comment (ekr) [inline draft-barnes-acme.md]: How can the client be providing the server's ECDH public key?

File: draft-barnes-acme.md [hunk 10]
```diff
@@ -516,18 +544,234 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
+  ],
+
+  "recoveryKey": {
+    "client": {
+      "kty": "EC",
+      "crv": "P-256",
+      "x": "...","y": "..."
+    },
```
Comment (ekr) [inline draft-barnes-acme.md]: Why does the server have to echo the client's key? It already knows it.

**Comment by ekr** [comment]
I have no idea what the intended security properties of this DH-based mechanism are, but this whole confirmation value thing seems quite dangerous.

Please write those down and then we can discuss what the right protocol mechanisms are.

File: draft-barnes-acme.md [hunk 11]
```diff
@@ -516,18 +544,234 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+    "length": 80,
+    "confirm": "_VPtXjAFDBC0O_JTnv8MCS7uTyb1NVlQtUvr5xcLjOo"
+  }
+}
+
+~~~~~~~~~~
+
+On receiving the server's response, the client MUST compute the recovery key and check that it matches the confirmation value.  (This ensures that the "length" field was not tampered with in transit.)  If the confirmation value does not match, then the client should discard the recovery key; it will not be usable for recovery on this registration.
+
```
Comment (ekr) [inline draft-barnes-acme.md]: Except that it doesn't demonstrate that the server's PK wasn't modified in transit. How are you in a worse situatio here?

File: draft-barnes-acme.md [hunk 12]
```diff
@@ -516,18 +544,234 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+  "resource": "reg",
+  "recoveryKey": {
+    "client": {
+      "kty": "EC",
+      "crv": "P-256",
+      "x": "...","y": "..."
+    }
+  }
+}
+/* Signed as JWS */
```
Comment (ekr) [inline draft-barnes-acme.md]: If this whole thing is signed, how does someone tamper with the length field?

**Comment by bifurcation** [comment]
Just to recap a phone call I had with @ekr and @martinthomson:

* The confirmation is necessary if the server sets the length, since the server response is not signed.
  * TODO: Have client send length, remove confirmation
  * TODO: Have server reject request if length is too short
* The server's response does not need to echo the client's key
  * TODO: Remove client key from server response
* There is a possibility of confusion between this usage of key agreement and the DNS challenge
  * TODO: Add a label to the KDF to distinguish

**Comment by bifurcation** [comment]
re-r? @ekr

File: draft-barnes-acme.md [hunk 13]
```diff
@@ -427,6 +435,34 @@ of {{RFC7515}}.  If the value of a "nonce" header parameter is not
...
+  * Client-generated key pair
+  * Server-generated key pair
+  * Length of the shared secret to be derived
+  * Label
+* Perform the ECDH primitive operation to obtain Z (Section 3.3.1 of {{SEC1}})
+* Select a hash algorithm according to the curve being used:
+  * For "P-256", use SHA-256
+  * For "P-384", use SHA-384
+  * For "P-512", use SHA-512
+* Derive the shared secret value using the KDF in Section 3.6.1 of {{SEC1}} using Z and the selected hash algorithm, and with the UTF-8 encoding of the label as the SharedInfo value
```
Comment (ekr) [inline draft-barnes-acme.md]: Is there any reason to tie this to the hash function as opposed to just SHA-256?

File: draft-barnes-acme.md [hunk 14]
```diff
@@ -427,6 +435,34 @@ of {{RFC7515}}.  If the value of a "nonce" header parameter is not
...
+
+~~~~~~~~~~
+H( Z || 00000001 || label )
+~~~~~~~~~~
+
+where:
+
+* H is the chosen hash algorithm
+* 00000001 is a four-octet value in hex
+* label is the UTF-8 value of the label
```
Comment (ekr) [inline draft-barnes-acme.md]: This is a weird way to write this, since you have the KDF above.

File: draft-barnes-acme.md [hunk 15]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
 
-Servers SHOULD NOT respond to GET requests for registration resources as these requests are not authenticated.  If a client wishes to query the server for information about its account (e.g., to examine the "contact" or "certificates" fields), then it SHOULD do so by sending a POST request with an empty update.  That is, it should send a JWS whose payload is the empty JSON dictionary ("{}").
+Servers SHOULD NOT respond to GET requests for registration resources as these requests are not authenticated.  If a client wishes to query the server for information about its account (e.g., to examine the "contact" or "certificates" fields), then it SHOULD do so by sending a POST request with an empty update.  That is, it should send a JWS whose payload is trivial ({"resource":"reg"}).
 
+### Recovery Keys
 
-### Account Recovery
+If the client wishes to establish a secret key with the server that it can use to recover this account later (a "recovery key"), then it must perform a simple key agreement protocol as part of the new-registration transaction.  The client and server perform an ECDH exchange through the new-registration transaction, and the result is the recovery key.
+
```
Comment (ekr) [inline draft-barnes-acme.md]: Cross-reference the above DH definition

File: draft-barnes-acme.md [hunk 16]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+To request a recovery key, the client includes a "recoveryKey" field in its new-registration request.  The value of this field is a JSON object.
+
+client (required, JWK):
+: The client's ECDH public key
+
+length (required, number):
+: The length, in bits, of the derived secret.  The value MUST be a multiple of 8 and SHOULD be at least 64.
```
Comment (ekr) [inline draft-barnes-acme.md]: Wh not use octets?

File: draft-barnes-acme.md [hunk 17]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+
+In the client's request, this object contains a JWK for a random ECDH public key generated by the client and the client-selected length value.  Clients SHOULD choose large length values, but may choose smaller ones for usability, e.g., in order to make it easy for a user to write down.
+
```
Comment (ekr) [inline draft-barnes-acme.md]: Can you describe somewhere what the threat model is? Is that below?

File: draft-barnes-acme.md [hunk 18]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+POST /acme/new-registration HTTP/1.1
+Host: example.com
+
+{
+  "resource": "new-reg",
+  "contact": [
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
+  ],
```
Comment (ekr) [inline draft-barnes-acme.md]: Why isn't this a dictionary?

File: draft-barnes-acme.md [hunk 19]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+  }
+}
+/* Signed as JWS */
+
+~~~~~~~~~~
+
+The server MUST validate that the elliptic curve ("crv") and length value chosen by the client are acceptable, and that is otherwise willing to create a recovery key.  If not, then it MUST reject the new-registration request.
+
+If the server agrees to create a recovery key, then it generates its own random ECDH key pair and combines it with with the client's public key as described in {{key-agreement}} above, using the label "recovery".  The derived secret value is the recovery key.  The server then returns to the client the ECDH key that it generated.
+
```
Comment (ekr) [inline draft-barnes-acme.md]: Does the ECDH key have to be fresh?

File: draft-barnes-acme.md [hunk 20]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+~~~~~~~~~~
+
+POST /acme/recover-registration HTTP/1.1
+Host: example.com
+
+{
+  "resource": "recover-reg",
+  "method": "token",
+  "base": "https://example.com/acme/reg/asdf",
```
Comment (ekr) [inline draft-barnes-acme.md]: Do we still want to use "token" ere?

File: draft-barnes-acme.md [hunk 21]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+/* Signed as JWS, with new account key */
+
+~~~~~~~~~~
+
+On receiving such a request the server MUST verify that:
+
+* The base registration has a recovery key associated with it
+* The "alg" value represents a MAC algorithm
+* The MAC verifies using the recovery key
+* The JWK in the payload represents the new account key (i.e. the key used to verify the ACME message)
```
Comment (ekr) [inline draft-barnes-acme.md]: s/verifies/matches/

File: draft-barnes-acme.md [hunk 22]
```diff
@@ -516,18 +554,214 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
...
+Link: <https://example.com/acme/new-authz>;rel="next"
+Link: <https://example.com/acme/recover-reg>;rel="recover"
+Link: <https://example.com/acme/terms>;rel="terms-of-service"
+
+{
+  "key": { /* JWK from JWS header */ },
+
+  "contact": [
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
```
Comment (ekr) [inline draft-barnes-acme.md]: Is this required?

File: draft-barnes-acme.md (hunk 13 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: Scaling the hash seemed to match other existing precedent, e.g., [Suite B](https://www.nsa.gov/ia/programs/suiteb_cryptography/):

> Curve P-256 for SECRET ... Curve P-384 for TOP SECRET ...
> SHA-256 for SECRET ... SHA-384 for TOP SECRET ...

Unless you have a reason *not* to scale the hash, I'm going to keep it as is.

It is 521.

File: draft-barnes-acme.md (hunk 14 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: It's just a summary of the KDF.  I will make this clearer.

File: draft-barnes-acme.md (hunk 16 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: Yeah, I went back and forth on this.  I went with bits because that seems like how people think about key sizes.  But I agree that the multiple of 8 thing looks silly; I might change it back.

File: draft-barnes-acme.md (hunk 17 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: I think the right place to handle that is in the security considerations, which I'm hoping to get started today.  So I'm planning to not do it in this PR.

File: draft-barnes-acme.md (hunk 18 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: The overall new-registration dict is described above; this is just describing the recoveryKey extension to it.  In particular, "contact" is defined above.

I could see an argument that this section could be folded in with the new-registration section, but given that it's fairly complex in itself, it makes sense to me to have it separated for now.  If we decide to fold them together, we can do it in a later PR.

File: draft-barnes-acme.md (hunk 19 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: As long as the client provides an ephemeral key, I think it's OK for the server's key to be constant across various transaction; the keys still turn out unique.  OTOH, it seems safer to be fresh.

File: draft-barnes-acme.md (hunk 20 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: Suggestion?  "authorization"?  Since the MAC asserts the new key under the authority of the recovery key holder.

File: draft-barnes-acme.md (hunk 21 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: Sorry, which "verify"?

File: draft-barnes-acme.md (hunk 22 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: You mean the "contact" field?  That's part of the registration object, so since we're giving the registration to this new key, we give it the object.

File: draft-barnes-acme.md (hunk 20 above)
Comment (bifurcation) [inline draft-barnes-acme.md]: Using "mac" because "MAC-based recovery" sounds better than "Authorization-based recovery"

File: draft-barnes-acme.md (hunk 19 above)
Comment (ekr) [inline draft-barnes-acme.md]: If you think it should be fresh, then i suggest normative language

File: draft-barnes-acme.md (hunk 21 above)
Comment (ekr) [inline draft-barnes-acme.md]: "The MAC verifies using the recovery key"

File: draft-barnes-acme.md (hunk 13 above)
Comment (ekr) [inline draft-barnes-acme.md]: Well, it's extra complexity, and it's hard to understand what the security benefit is.

Incidentally, how does the client know which curves the server supports? What happens if the client proposes one the server doesn't like.

**Comment by bifurcation** [comment]
(marking r=ekr with his out-of-band permission)
